src/coreclr/src/jit/rangecheck.cpp

```diff
@@ -773,11 +773,53 @@ void RangeCheck::MergeAssertion(BasicBlock* block, GenTree* op, Range* pRange DE
 // Compute the range for a binary operation.
 Range RangeCheck::ComputeRangeForBinOp(BasicBlock* block, GenTreeOp* binop, bool monIncreasing DEBUGARG(int indent))
 {
-    assert(binop->OperIs(GT_ADD));
+    assert(binop->OperIs(GT_ADD, GT_AND, GT_RSH, GT_LSH, GT_UMOD));
 
     GenTree* op1 = binop->gtGetOp1();
     GenTree* op2 = binop->gtGetOp2();
 
+    if (binop->OperIs(GT_AND, GT_RSH, GT_LSH, GT_UMOD))
+    {
+        if (!op2->IsIntCnsFitsInI32())
+        {
+            // only cns is supported for op2 at the moment for &,%,>> operators
+            return Range(Limit::keUnknown);
+        }
+
+        int icon = -1;
+        if (binop->OperIs(GT_AND))
+        {
+            // x & cns -> [0..cns]
+            icon = static_cast<int>(op2->AsIntCon()->IconValue());
+        }
+        else if (binop->OperIs(GT_UMOD))
+        {
+            // x % cns -> [0..cns-1]
+            icon = static_cast<int>(op2->AsIntCon()->IconValue()) - 1;
+        }
+        else if (binop->OperIs(GT_RSH, GT_LSH) && op1->OperIs(GT_AND) && op1->AsOp()->gtGetOp2()->IsIntCnsFitsInI32())
+        {
+            // (x & cns1) >> cns2 -> [0..cns1>>cns2]
+            int icon1 = static_cast<int>(op1->AsOp()->gtGetOp2()->AsIntCon()->IconValue());
+            int icon2 = static_cast<int>(op2->AsIntCon()->IconValue());
+            if (icon1 >= 0 && icon2 >= 0 && icon2 < 32)
```

Comment:
Subexpressions have to be in parentheses: https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/clr-jit-coding-conventions.md#111-logical-and-arithmetic-expressions